src/EditorFeatures/CSharp/CompleteStatement/CompleteStatementCommandHandler.cs

```diff
@@ -0,0 +1,397 @@
+
+using System;
+using System.ComponentModel.Composition;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Extensions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Editor.Implementation.AutomaticCompletion;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Host.Mef;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Commanding;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor.Commanding.Commands;
+using Microsoft.VisualStudio.Text.Operations;
+using Microsoft.VisualStudio.Utilities;
+using VSCommanding = Microsoft.VisualStudio.Commanding;
+
+namespace Microsoft.CodeAnalysis.Editor.CSharp.CompleteStatement
+{
+    /// <summary>
+    /// When user types <c>;</c> in a statement, closing delimiters and semicolon are added and caret is placed after the semicolon
+    /// </summary>
+    [Export(typeof(VSCommanding.ICommandHandler))]
+    [ContentType(ContentTypeNames.CSharpContentType)]
+    [Name(nameof(CompleteStatementCommandHandler))]
+    [Order(After = PredefinedCommandHandlerNames.Completion)]
+    internal sealed class CompleteStatementCommandHandler : IChainedCommandHandler<TypeCharCommandArgs>
+    {
+        private readonly ITextUndoHistoryRegistry _undoHistoryRegistry;
+        private readonly IEditorOperationsFactoryService _editorOperationsFactoryService;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public CompleteStatementCommandHandler(ITextUndoHistoryRegistry undoHistoryRegistry, IEditorOperationsFactoryService editorOperationsFactoryService)
+        {
+            _undoHistoryRegistry = undoHistoryRegistry;
+            _editorOperationsFactoryService = editorOperationsFactoryService;
+        }
+
+        public string DisplayName => CSharpEditorResources.Complete_statement_on_semicolon;
+
+        public void ExecuteCommand(TypeCharCommandArgs args, Action nextCommandHandler, CommandExecutionContext executionContext)
+        {
+            BeforeExecuteCommand(args, executionContext);
+            nextCommandHandler();
+        }
+
+        private void BeforeExecuteCommand(TypeCharCommandArgs args, CommandExecutionContext executionContext)
+        {
+            if (args.TypedChar != ';')
+            {
+                return;
+            }
+
+            var caret = args.TextView.GetCaretPoint(args.SubjectBuffer);
+            if (!caret.HasValue)
+            {
+                return;
+            }
+
+            var isCaretAtEndOfLine = ((SnapshotPoint)caret).Position == ((SnapshotPoint)caret).GetContainingLine().End;
```

Comment:
instead of casting CaretPosition multiple times, just make a local and use that.

Reply:
this value should be computed nearer to where it is needed.

Reply:
is it necessary to truely be at the exact end of line?  what about if there is trailing whitespace?  it seems weird to me that we would treaat those cases differently.

Reply:
> instead of casting CaretPosition multiple times, just make a local and use that.

Duplicate of https://github.com/dotnet/roslyn/pull/30743#discussion_r242668495

Reply:
When getting the node at the current caret position, it typically returns the node associated with the space immediately following the caret. However if the caret is at the end of the line, it returns the node of the space preceding it.  So in this, `string s = "Hello"$$`, the SyntaxKind of the node at that caret position is SyntaxKind.StringLiteralExpression and normally would not be eligible for statement completion.  But since the cursor is actually outside the string, it should be eligible.